[nexus] add generate_kspace_jastrow

Paul-St-Young opened the pull request:
Added `generate_kspace_jastrow` in the same style as `generate_jastrow1,2,3`.

I want to add a default long-range Jastrow to `generate_jastrow`, but I need help with setting reasonable defaults. I think a safe default is: no one-body part and RPA two-body part. This means `kc1=0`, `nk1=0`, `kc2=1~2*kf`, `nk2=[# of kshells inside kc2]`.

To enable this default, I need access to Fermi kvector magnitude `kf` and count the number of kshells inside `kf`. Some routines in `structure.py` (designed for tiling), can be used for this purpose. I am tempted to add `calc_kf` and `count_nsh` to `structure.py`, but I am not sure if this is the best route. @jtkrogel what do you think?

jtkrogel commented:
Adding functions to structure.py is fine.  Names like `kf_rpa` and `count_kshells` would be more descriptive IMO.

jtkrogel commented on nexus/lib/qmcpack_input.py:
```diff
@@ -5254,6 +5255,57 @@ def generate_jastrow3(function='polynomial',esize=3,isize=3,rcut=4.,coeff=None,i
+  if coeff1 is None: coeff1 = [0]*nk1
+  if coeff2 is None: coeff2 = [0]*nk2
+  if len(coeff1) != nk1: raise RuntimeError('coeff1 mismatch')
+  if len(coeff2) != nk2: raise RuntimeError('coeff2 mismatch')
```
Please use the standard error handling mechanisms in Nexus.  In this case, `QmcpackInput.class_error('coeff2 mismatch','generate_kspace_jastrow')` would do fine.  These errors can all be converted to raise type errors under the hood (as nxs-test does) while defaulting to "abort with traceback" for typical use.

jtkrogel commented on nexus/lib/qmcpack_input.py:
```diff
@@ -5254,6 +5255,57 @@ def generate_jastrow3(function='polynomial',esize=3,isize=3,rcut=4.,coeff=None,i
+      'id': 'cG1', 'type': 'Array',
+      'coeff': coeff1
+     })
+  })
```
This looks a little heavy syntactically.  Does this not work:

  corr1 = correlation(
    type = 'One-Body',
    symmetry = symm1,
    kc = kc1,
    coefficients = section(
      id = 'cG1', type = 'Array',
      coeff = coeff1
     )
  )

I generally prefer this form for Nexus since it removes two layers of markup: one for the dict "{}" and one for each key "''".

Paul-St-Young replied:
I use dictionary in my own scripts because then I can pass the dictionary around to other simple scripts that know nothing about nexus. Now, since this function is being added to nexus, I agree that the internal syntax is cleaner.

Paul-St-Young commented:
@jtkrogel I address your concerns and added the default k-space jastrow to `generate_jastrows`. Please see if this PR is good to merge now.

I added a few unit tests to the folder `nexus/tests/unit` in the style of pytest/nosetest. Is this OK with you?

prckent commented:
@jtkrogel  Please can you mark this as changes requested if that is still the situation.

jtkrogel commented:
Sorry, I'm being a little slow getting to this.

jtkrogel commented on nexus/lib/physical_system.py:
```diff
@@ -635,6 +635,15 @@ def ae_pp_species(self):
     #end def ae_pp_species
+
+    def kf_rpa(self):
+      import numpy as np
```
Move this to a top-level import.  Nexus will move to this standard pattern over time, but no need for local import.

jtkrogel commented on nexus/lib/physical_system.py:
```diff
@@ -635,6 +635,15 @@ def ae_pp_species(self):
+      nelecs = self.particles.electron_counts()
+      volume = self.structure.volume()
+      kvol1 = (2*np.pi)**3/volume  # k-space volume per particle
+      kfs = [(3*nelec*kvol1/(4*np.pi))**(1./3) for nelec in nelecs]
```
Prefer ndarray return type.

jtkrogel commented on nexus/lib/qmcpack_input.py:
```diff
@@ -4786,6 +4787,11 @@ def generate_jastrows(jastrows,system=None,return_list=False,check_ions=False):
+        if 'k' in jorders:
+            kcut = max(system.rpa_kf())
+            nksh = system.structure.count_kshells(kcut)
+            jterm = generate_kspace_jastrow(0, kcut, 0, nksh)
```
Use keyword inputs where possible as these break less easily under future changes than positional arguments do.

jtkrogel commented on nexus/lib/qmcpack_input.py:
```diff
@@ -5254,6 +5260,59 @@ def generate_jastrow3(function='polynomial',esize=3,isize=3,rcut=4.,coeff=None,i
+  symm1='isotropic', symm2='isotropic', coeff1=None, coeff2=None):
+  """ generate <jastrow type="kSpace">
+
+  Args:
```
Move to NumPy style docstrings.

jtkrogel commented on nexus/lib/qmcpack_input.py:
```diff
@@ -5254,6 +5260,59 @@ def generate_jastrow3(function='polynomial',esize=3,isize=3,rcut=4.,coeff=None,i
+  if coeff1 is None: coeff1 = [0]*nk1
+  if coeff2 is None: coeff2 = [0]*nk2
+  if len(coeff1) != nk1:
+    QmcpackInput.class_error('coeff1 mismatch', 'generate_kspace_jastrow')
```
Add `#end if` closures.

jtkrogel commented on nexus/lib/structure.py:
```diff
@@ -1223,6 +1223,36 @@ def distances(self,pos1=None,pos2=None):
+        #end if
+      #end for
+
+      import numpy as np
```
Move to top level import.

jtkrogel commented on nexus/tests/unit/test_qmcpack_input.py:
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env python
+def test_generate_kspace_jastrow():
```
Similarly repackage into nxs-test with the contents of "main" and all other local functions going into a new function called "qmcpack_input".

jtkrogel commented:
@Paul-St-Young Thanks for your patience.  You caught me in the middle of settling on docstring styles.

Most of the remaining requested changes are small.  Hopefully repackaging the tests will not take too much effort.

Your inclusion of documentation and tests are greatly appreciated.

Paul-St-Young commented:
@jtkrogel sorry I did not see your comments until now. I addressed most of your comments (everything except for the unit tests).

I am a little intimidated by the size of `nxs-test`. Do you think it makes sense to include only the difficult tests (like job monitoring and integration tests) in `nxs-test` and split out simple unit tests of stand-alone objects like `strcuture` and `qmcpack_input` in standard pytest/nosetest style?

I prefer the latter because:
1. Many users can contribute tests on different objects simultaneously without merge conflict.
2. New users need not learn the internal syntax of `nxs-test`, only the standard pytest/nosetest syntax.
3. This helps make `nxs-test` less intimidating to edit.

Please let me know if this is OK with you (i.e. keep the unit tests as is).
If you still prefer keeping all the tests in one file, then I will repackage the tests tomorrow (hopefully...)

jtkrogel commented on nexus/tests/unit/test_physical_system.py:
```diff
@@ -0,0 +1,43 @@
+  assert np.isclose(kfs[1], 1.465/2**(1./3), atol=1e-3)
+#end def test_kf_rpa
+
+def test_count_kshells():
```
Move `test_count_kshells` into a new file named `test_structure.py` as this tests a Structure object rather than a PhysicalSystem object.

jtkrogel commented:
@Paul-St-Young nxs-test needs to be able to run all the tests.  I agree that supporting pytest style development is also desirable.  

I have a solution in the works that will allow easy extension of nxs-test to include tests developed in standard pytest style that I will put in a PR after this one goes in.  This PR can go in once the tests are split up as requested above.

jtkrogel commented:
NB: PR #1765 is the PR referred to above.

Paul-St-Young commented:
@jtkrogel I split out `test_count_kshells` as you requested.

Thanks for being accommodating to these different styles!

jtkrogel commented:
Thanks, and you're welcome.  Enabling standard style unit tests should increase the number of contributed tests, which is highly useful.

prckent commented:
Test this please

prckent commented:
OK to test
